source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,70 @@
+option to enable them for the given cluster.
+
+TODO: Description of the UI actions once they are implemented.
+
+## Changes to Engine and Vdsm
+
+There are no special changes needed in order to add a TPM device as such, it's just another kind of device added to Engine.
+
+But a very different thing is ensuring TPM data persistence.  TPM data is managed using an external tool, swtpm.  Once the VM domain XML is deleted on the host, the TPM data is deleted as well by libvirt and it must be restored before the VM is run again, whether on this or another host (live host migrations are handled transparently and don't require any special actions on the oVirt side).
+
+In order to support TPM data persistence, we must store the data in the Engine database and transfer it to the host before the VM is started and update it in the Engine database after the VM gets down.  We need the following Vdsm API changes to achieve that:
+
+- Adding a new VM.create parameter providing the data.
+- Adding a new parameter to VM.destroy call response containing the updated data.
+
+There is some fragility with that approach.  If VM.destroy call fails for any reason after the libvirt domain is undefined (and the corresponding TPM data is deleted by libvirt), the updated data will be lost.  But we don't have many alternatives: While we could make a separate API call to retrieve TPM data or attach them to VM stats, that would mean retrieving live data from swtpm, which may not be consistent.  Only data retrieved after the VM stops can be considered stable.  And even with a separate API call, if Engine loses contact with the host and starts the VM elsewhere, updates to TPM data since the VM start will be lost.  We cannot guarantee, without introducing complicated implementations, up-to-date TPM data if the environment doesn't work as expected.  Considering TPM data typically serves forensic purposes rather than storing critical information, we may accept the fact that VM guest OS repair is needed in case the data got changed and the environment experienced unexpected trouble (i.e. VM.destroy call hasn't been successful).  It may still cause trouble for HA VMs.
+
+An alternative mechanism could be to use shared storage to store TPM data, either by copying the data there or (preferably, if possible) to let swtpm use the shared storage directly.  The differences to the solution proposed above would be:
+
+- More complicated implementation.
+- Need to connect stored data to the corresponding (disk-only) snapshots, either on the storage, or by copying them to and from the database.
+- Better data safety in case of direct use of the shared storage by swtpm, the data is updated immediately, not only once VM stops running (but there may in theory data consistency problems again).
+- It's unclear how swtpm would work with network storage and how to prevent libvirt from deleting the data once the domain XML is deleted on the host.
```

Comment:
That's an interesting approach.
Maybe we can combine the two together? I mean, to let VDSM saves the data to those volumes and read from them when starting the VMs instead of sending the data to the engine?

Reply:
I discussed the idea of putting the data to storage with Nir and it doesn't look like a very simple thing. A volume would be needed with all the related arrangements and it needn't be absolutely reliable too -- we know storage problems can occur for many reasons. We came to the conclusion that a better idea could be to make a special Vdsm API call for retrieving TPM data that would be called periodically. Engine could ask for the data with a relatively long interval or whenever TPM data checksum reported by Vdsm changes (what would be the complications you mentioned above with storing VM data hashes on the Engine side and acting when they change?). The last call would be performed just before calling the final Destroy (to avoid problems with failures after the data is wiped on the host).
This way, Engine would be managing the data transfers, based on some information from Vdsm, which looks more robust than other solutions. Are there any obstacles on the Engine side to implement something like that?

Reply:
I think that storing this data (in vm_dynamic) and sending it on RunVm should be fairly simple. The complication I thought of was more about a mechanism to get the hash and detecting a change. We do that for the 'devices hash' - we have VmDevicesMonitoring that on engine startup loads the hash of each VM from the database and then stores that in-memory and compares it with the one we get from the stats call. It's not rocket science to introduce another hash that this module or an additional module would handle, but it can be simpler to use the same 'devices hash' for that as well if possible (that hash already covers some of the guest agent data afair).

What I like about storing this data on the storage though is that it sounds more aligned with our desire to be able to (re)start VMs without the engine being involved. I guess we wouldn't want to store such sensitive data in the metadata, right?

Reply:
> What I like about storing this data on the storage though is that it sounds more aligned with our desire to be able to (re)start VMs without the engine being involved. 

That's a good remark and an argument important enough to consider using storage.

> I guess we wouldn't want to store such sensitive data in the metadata, right?

The data is already accessible to anybody with root access to the host and also, in most of the discussed implementation scenarios, to Vdsm. So I think having them in metadata is no big additional risk. But the question is whether we want to store larger data in metadata. While TPM capacity is AFAIK very limited (I'm not sure about emulated TPM though), it's still going to be at least 4 KB of encoded data with swtpm.

Still, using metadata would simplify things a lot. When using storage, we would have to arrange the volume, to ensure data can't be left in inconsistent state there (e.g. when the storage gets suddenly disconnected while TPM data is being written), and to deliver (consistent) data there, either by arranging swtpm to use the storage directly or making Vdsm copy it there.